File: internal/mods/modinfo_fallback.py
import json
import math
import os

from lupa import LuaRuntime, lua_type


modinfo_path = os.environ["DST_MODINFO_PATH"]
mod_id = os.environ.get("DST_MODINFO_ID", "")
selected_locale = os.environ.get("DST_MODINFO_LOCALE", "zh")

lua = LuaRuntime(unpack_returned_tuples=True)
globals_table = lua.globals()
original_globals = {str(key) for key in globals_table}

globals_table.locale = selected_locale
globals_table.folder_name = "workshop-" + mod_id
lua.execute(
    "ChooseTranslationTable = function(translations) "
    "return translations[locale] or translations.zhr or translations[1] end"
)
globals_table.modimport = globals_table.require
globals_table.print = lambda *args: None

lua_path = os.environ.get("LUA_PATH")
lua_cpath = os.environ.get("LUA_CPATH")
if lua_path:
    globals_table.package.path = lua_path.replace(";;", ";" + str(globals_table.package.path) + ";", 1)
if lua_cpath:
    globals_table.package.cpath = lua_cpath.replace(";;", ";" + str(globals_table.package.cpath) + ";", 1)

with open(modinfo_path, "rb") as source_file:
    source = source_file.read().decode("utf-8-sig")
lua.execute(source)


def convert(value, visiting, depth):
    value_type = lua_type(value)
    if value_type == "function":
        return None
    if value_type != "table":
        if isinstance(value, float) and not math.isfinite(value):
            return None
        return value
    if depth > 100:
        raise ValueError("Lua table nesting exceeds 100 levels")
    identity = id(value)
    if identity in visiting:
        raise ValueError("cyclic Lua table")
    visiting.add(identity)
    try:
        keys = list(value.keys())
        if all(isinstance(key, int) and not isinstance(key, bool) for key in keys):
            ordered = sorted(keys)
            if ordered == list(range(1, len(ordered) + 1)):
                return [convert(value[index], visiting, depth + 1) for index in ordered]
        result = {}
        for key in keys:
            if isinstance(key, bool):
                mapped_key = "true" if key else "false"
            elif isinstance(key, (str, int)):
                mapped_key = str(key)
            elif isinstance(key, float) and math.isfinite(key):
                mapped_key = str(key)
            else:
                raise TypeError("unsupported Lua table key type")
            if mapped_key in result:
                raise ValueError("duplicate JSON key after Lua key conversion: " + mapped_key)
            result[mapped_key] = convert(value[key], visiting, depth + 1)
        return result
    finally:
        visiting.remove(identity)


excluded = {"locale", "folder_name", "ChooseTranslationTable", "modimport"}
result = {}
for key in globals_table:
    name = str(key)
    value = globals_table[key]
    if name not in original_globals and name not in excluded and lua_type(value) != "function":
        result[name] = convert(value, set(), 0)

print(json.dumps(result, ensure_ascii=False, allow_nan=False, separators=(",", ":")))

File: internal/mods/modinfo_fallback.lua
local modinfo_path = assert(os.getenv("DST_MODINFO_PATH"), "DST_MODINFO_PATH is required")
local mod_id = os.getenv("DST_MODINFO_ID") or ""
local selected_locale = os.getenv("DST_MODINFO_LOCALE") or "zh"

local original_globals = {}
for key in pairs(_G) do
    original_globals[key] = true
end

locale = selected_locale
folder_name = "workshop-" .. mod_id
ChooseTranslationTable = function(translations)
    return translations[selected_locale] or translations.zhr or translations[1]
end
modimport = require
print = function() end

local chunk, load_error = loadfile(modinfo_path)
if not chunk then
    error(load_error)
end
chunk()

local escapes = {
    ['"'] = '\\"',
    ['\\'] = '\\\\',
    ['\b'] = '\\b',
    ['\f'] = '\\f',
    ['\n'] = '\\n',
    ['\r'] = '\\r',
    ['\t'] = '\\t',
}

local function encode_string(value)
    return '"' .. value:gsub('[%z\1-\31\\"]', function(character)
        return escapes[character] or string.format('\\u%04x', string.byte(character))
    end) .. '"'
end

local function table_shape(value)
    local count = 0
    local maximum = 0
    for key in pairs(value) do
        if type(key) ~= "number" or key < 1 or key ~= math.floor(key) then
            return false, 0
        end
        count = count + 1
        if key > maximum then
            maximum = key
        end
    end
    return count == maximum, maximum
end

local function encode(value, visiting, depth)
    local kind = type(value)
    if kind == "nil" then
        return "null"
    elseif kind == "boolean" then
        return value and "true" or "false"
    elseif kind == "number" then
        if value ~= value or value == math.huge or value == -math.huge then
            return "null"
        end
        return tostring(value)
    elseif kind == "string" then
        return encode_string(value)
    elseif kind == "function" then
        return "null"
    elseif kind ~= "table" then
        error("unsupported Lua value type: " .. kind)
    elseif depth > 100 then
        error("Lua table nesting exceeds 100 levels")
    end

    if visiting[value] then
        error("cyclic Lua table")
    end
    visiting[value] = true

    local parts = {}
    local is_array, length = table_shape(value)
    if is_array then
        for index = 1, length do
            parts[#parts + 1] = encode(value[index], visiting, depth + 1)
        end
        visiting[value] = nil
        return "[" .. table.concat(parts, ",") .. "]"
    end

    local mapped = {}
    for key, item in pairs(value) do
        local key_kind = type(key)
        if key_kind == "string" or key_kind == "number" or key_kind == "boolean" then
            if key_kind == "number" and (key ~= key or key == math.huge or key == -math.huge) then
                error("unsupported non-finite Lua table key")
            end
            local mapped_key = tostring(key)
            if mapped[mapped_key] ~= nil then
                error("duplicate JSON key after Lua key conversion: " .. mapped_key)
            end
            mapped[mapped_key] = item
        else
            error("unsupported Lua table key type: " .. key_kind)
        end
    end
    local keys = {}
    for key in pairs(mapped) do
        keys[#keys + 1] = key
    end
    table.sort(keys)
    for _, key in ipairs(keys) do
        parts[#parts + 1] = encode_string(key) .. ":" .. encode(mapped[key], visiting, depth + 1)
    end
    visiting[value] = nil
    return "{" .. table.concat(parts, ",") .. "}"
end

local excluded = {
    locale = true,
    folder_name = true,
    ChooseTranslationTable = true,
    modimport = true,
}
local result = {}
for key, value in pairs(_G) do
    if type(key) == "string" and not original_globals[key] and not excluded[key] and type(value) ~= "function" then
        result[key] = value
    end
end

io.write(encode(result, {}, 0))
